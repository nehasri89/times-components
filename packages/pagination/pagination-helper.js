import React from "react";
import PropTypes from "prop-types";

export default Component => {
  class Helper extends React.Component {
    constructor(props) {
      super(props);
      this.handleChangePage = this.handleChangePage.bind(this);
      this.state = props;
    }

    componentWillReceiveProps(nextProps) {
      return this.setState(
        Object.assign({}, nextProps, {
          page: this.state.page
        })
      );
    }

    handleChangePage(page) {
      return this.setState({ page });
    }

    render() {
      return (
        <Component
          {...this.state}
          onNext={this.handleChangePage}
          onPrev={this.handleChangePage}
        />
      );
    }
  }

  Helper.displayName = `Pagination Helper (${Component.displayName})`;
  Helper.propTypes = {
    page: PropTypes.number
  };

  Helper.defaultProps = {
    page: 1
  };

  return Helper;
};
