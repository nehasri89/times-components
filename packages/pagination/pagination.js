import React from "react";
import { StyleSheet, Text, TouchableHighlight, View } from "react-native";
import PropTypes from "prop-types";
import pagination from "./pagination-helper";

const styles = StyleSheet.create({
  compact: {
    position: "absolute",
    right: 0,
    left: 0
  },
  container: {
    alignItems: "stretch",
    display: "flex",
    flexDirection: "column"
  },
  horizontal: {
    display: "flex",
    flexDirection: "row",
    borderBottomColor: "#dbdbdb",
    borderBottomWidth: 1,
    borderStyle: "solid",
    borderTopColor: "#dbdbdb",
    borderTopWidth: 1,
    justifyContent: "space-between"
  },
  arrows: {
    color: "#006699",
    fontFamily: "GillSansMTStd-Medium",
    fontSize: 14,
    lineHeight: 38
  },
  label: {
    color: "#696969",
    fontFamily: "GillSansMTStd-Medium",
    fontSize: 15,
    lineHeight: 20,
    textAlign: "center",
    paddingBottom: 11,
    paddingTop: 11
  }
});

const shouldRenderOneLine = width => width >= 700;

class Pagination extends React.Component {
  constructor(props) {
    super(props);

    const {
      compact,
      count,
      generatePageLink,
      page,
      pageSize,
      onNext,
      onPrev
    } = props;

    this.state = {
      compact,
      count,
      isCompact: false,
      generatePageLink,
      page,
      pageSize,
      onNext,
      onPrev
    };

    this.handleLayout = this.handleLayout.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    this.setState(
      Object.assign({}, nextProps, {
        isCompact: this.state.isCompact
      })
    );
  }

  handleLayout({ nativeEvent }) {
    const { width } = nativeEvent.layout;

    return this.setState({
      isCompact: shouldRenderOneLine(width)
    });
  }

  render() {
    const {
      compact,
      count,
      generatePageLink,
      page,
      pageSize,
      onNext,
      onPrev
    } = this.state;

    const startResult = (page - 1) * pageSize + 1;
    const finalResult = Math.min(count, page * pageSize);
    const message = `Showing ${startResult} - ${finalResult} of ${count} results`;

    const prevComponent = startResult > pageSize
      ? <TouchableHighlight>
          <Text
            accessibilityRole="link"
            href={generatePageLink(page - 1)}
            style={styles.arrows}
            onPress={() => onPrev(page - 1)}
          >
            {"< Previous page"}
          </Text>
        </TouchableHighlight>
      : null;

    const nextComponent = finalResult < count
      ? <TouchableHighlight>
          <Text
            accessibilityRole="link"
            href={generatePageLink(page + 1)}
            style={styles.arrows}
            onPress={() => onNext(page + 1)}
          >
            {"Next page >"}
          </Text>
        </TouchableHighlight>
      : null;

    const messageComponent = !compact
      ? <Text
          style={[styles.label, this.state.isCompact ? styles.compact : null]}
        >
          {message}
        </Text>
      : null;

    return (
      <View style={styles.container} onLayout={this.handleLayout}>
        {messageComponent}
        <View style={styles.horizontal}>
          <View>{prevComponent}</View>
          <View>{nextComponent}</View>
        </View>
      </View>
    );
  }
}

Pagination.propTypes = {
  compact: PropTypes.bool,
  count: PropTypes.number.isRequired,
  page: PropTypes.number,
  pageSize: PropTypes.number,
  generatePageLink: PropTypes.func,
  onNext: PropTypes.func,
  onPrev: PropTypes.func
};

Pagination.defaultProps = {
  compact: false,
  generatePageLink: page => `./${page}`,
  page: 1,
  pageSize: 20,
  onNext: () => {},
  onPrev: () => {}
};

export default Pagination;

export { pagination };
