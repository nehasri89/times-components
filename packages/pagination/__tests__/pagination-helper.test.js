/* eslint-env jest */

import React from "react";
import renderer from "react-test-renderer";
import { Text } from "react-native";
import { shallow } from "enzyme";
import pagination from "../pagination-helper";

it("renders inner component with page 1", () => {
  const Component = props => <Text>{JSON.stringify(props, null, 2)}</Text>;
  const PageChanger = pagination(Component);

  const props = {
    page: 1
  };

  const component = renderer.create(<PageChanger {...props} />);
  expect(component).toMatchSnapshot();
});

it("renders inner component with new props", () => {
  const Component = props => <Text>{JSON.stringify(props, null, 2)}</Text>;
  const PageChanger = pagination(Component);

  const props = {
    foo: "not bar",
    page: 1
  };

  const wrapper = shallow(<PageChanger {...props} />);

  wrapper.setProps({ foo: "bar" });
  wrapper.update();

  expect(wrapper.state().foo).toEqual("bar");
  expect(wrapper.state().page).toEqual(1);
  expect(wrapper).toMatchSnapshot();
});

it("renders inner component with prev page", () => {
  const Component = props => <Text>{JSON.stringify(props, null, 2)}</Text>;
  const PageChanger = pagination(Component);

  const props = {
    page: 2
  };

  const wrapper = shallow(<PageChanger {...props} />);
  wrapper.instance().handleChangePage(1);
  wrapper.update();

  expect(wrapper.state().page).toEqual(1);
  expect(wrapper).toMatchSnapshot();
});

it("renders inner component with next page", () => {
  const Component = props => <Text>{JSON.stringify(props, null, 2)}</Text>;
  const PageChanger = pagination(Component);

  const props = {
    page: 2
  };

  const wrapper = shallow(<PageChanger {...props} />);
  wrapper.instance().handleChangePage(3);
  wrapper.update();

  expect(wrapper.state().page).toEqual(3);
  expect(wrapper).toMatchSnapshot();
});
