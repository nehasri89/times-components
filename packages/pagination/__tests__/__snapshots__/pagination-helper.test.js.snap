// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders inner component with new props 1`] = `
<Component
  foo="bar"
  onNext={[Function]}
  onPrev={[Function]}
  page={1}
/>
`;

exports[`renders inner component with next page 1`] = `
<Component
  onNext={[Function]}
  onPrev={[Function]}
  page={3}
/>
`;

exports[`renders inner component with page 1 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  {
    "page": 1
  }
</Text>
`;

exports[`renders inner component with prev page 1`] = `
<Component
  onNext={[Function]}
  onPrev={[Function]}
  page={1}
/>
`;
